src/common/security/admiral/context.go

```diff
@@ -71,74 +71,27 @@ func (s *SecurityContext) IsSolutionUser() bool {
 
 // HasReadPerm returns whether the user has read permission to the project
 func (s *SecurityContext) HasReadPerm(projectIDOrName interface{}) bool {
-	public, err := s.pm.IsPublic(projectIDOrName)
-	if err != nil {
-		log.Errorf("failed to check the public of project %v: %v",
-			projectIDOrName, err)
-		return false
-	}
-	if public {
-		return true
-	}
-
-	// private project
-	if !s.IsAuthenticated() {
-		return false
-	}
-
-	// system admin
-	if s.IsSysAdmin() {
-		return true
-	}
-
-	roles := s.GetProjectRoles(projectIDOrName)
-
-	return len(roles) > 0
+	return s.HasProjectPerm(projectIDOrName, ram.Resource("image"), ram.Action("pull"))
 }
 
 // HasWritePerm returns whether the user has write permission to the project
 func (s *SecurityContext) HasWritePerm(projectIDOrName interface{}) bool {
-	if !s.IsAuthenticated() {
-		return false
-	}
-
-	// system admin
-	if s.IsSysAdmin() {
-		return true
-	}
-
-	roles := s.GetProjectRoles(projectIDOrName)
-	for _, role := range roles {
-		switch role {
-		case common.RoleProjectAdmin,
-			common.RoleDeveloper:
-			return true
-		}
-	}
-
-	return false
+	return s.HasProjectPerm(projectIDOrName, ram.Resource("image"), ram.Action("push"))
 }
 
 // HasAllPerm returns whether the user has all permissions to the project
 func (s *SecurityContext) HasAllPerm(projectIDOrName interface{}) bool {
-	if !s.IsAuthenticated() {
-		return false
-	}
-
-	// system admin
-	if s.IsSysAdmin() {
-		return true
-	}
+	return s.HasProjectPerm(projectIDOrName, ram.Resource("image"), ram.Action("all"))
+}
 
-	roles := s.GetProjectRoles(projectIDOrName)
-	for _, role := range roles {
-		switch role {
-		case common.RoleProjectAdmin:
-			return true
-		}
-	}
+// HasProjectPerm returns whether the user has action permission of resource to the project
+func (s *SecurityContext) HasProjectPerm(projectIDOrName interface{}, subresource ram.Resource, action ram.Action) bool {
```

Comment:
It seems this func is unnecessary?  Or will there be another PR to remove these funcs from the interface and implementations?

Reply:
Will be removed in the next PR. If necessary, i will remove it in this pr.

Reply:
HasProjectPrem removed